transport/src/main/java/io/netty/channel/pool/FixedChannelPool.java

```diff
@@ -0,0 +1,354 @@
+                acquire0(promise);
+            } else {
+                executor.execute(new OneTimeTask() {
+                    @Override
+                    public void run() {
+                        acquire0(promise);
+                    }
+                });
+            }
+        } catch (Throwable cause) {
+            promise.setFailure(cause);
+        }
+        return promise;
+    }
+
+    private void acquire0(final Promise<Channel> promise) {
+        assert executor.inEventLoop();
+
+        if (acquiredChannelCount < maxConnections) {
+            ++acquiredChannelCount;
+
+            assert acquiredChannelCount > 0;
+
+            // We need to create a new promise as we need to ensure the AcquireListener runs in the correct
+            // EventLoop
+            Promise<Channel> p = executor.newPromise();
+            p.addListener(new AcquireListener(promise));
+            super.acquire(p);
+        } else {
+            if (pendingAcquireCount >= maxPendingAcquires) {
+                promise.setFailure(FULL_EXCEPTION);
+            } else {
+                AcquireTask task = new AcquireTask(promise);
+                if (pendingAcquireQueue.offer(task)) {
+                    ++pendingAcquireCount;
+
+                    if (timeoutTask != null) {
+                        task.timeoutFuture = executor.schedule(timeoutTask, acquireTimeoutNanos, TimeUnit.NANOSECONDS);
+                    }
+                } else {
+                    promise.setFailure(FULL_EXCEPTION);
+                }
+            }
+
+            assert pendingAcquireCount > 0;
+        }
+    }
+
+    @Override
+    public Future<Void> release(final Channel channel, final Promise<Void> promise) {
+        try {
+            final Promise<Void> p = executor.newPromise();
+            super.release(channel, p.addListener(new FutureListener<Void>() {
+
+                @Override
+                public void operationComplete(Future<Void> future) throws Exception {
+                    assert executor.inEventLoop();
+
+                    if (future.isSuccess()) {
+                        --acquiredChannelCount;
+
+                        // We should never have a negative value.
+                        assert acquiredChannelCount >= 0;
+
+                        // Run the pending acquire tasks before notify the original promise so if the user would
+                        // try to acquire again from the ChannelFutureListener and the pendingAcquireCount is >=
+                        // maxPendingAcquires we may be able to run some pending tasks first and so allow to add
+                        // more.
+                        runTaskQueue();
+                        promise.setSuccess(null);
+                    } else {
+                        promise.setFailure(future.cause());
+                    }
+                }
+            }));
+            return p;
+        } catch (Throwable cause) {
+            promise.setFailure(cause);
```

Comment:
What should with our state if an exception is thrown?  Should we decrement our acquired count, `runTaskQueue()`, etc...?